x-pack/plugin/core/src/main/resources/monitoring-apm-server.json

```diff
@@ -0,0 +1,686 @@
+            "beat": {
+              "properties": {
+                "host": {
+                  "type": "keyword"
+                },
+                "name": {
+                  "type": "keyword"
+                },
+                "type": {
+                  "type": "keyword"
+                },
+                "uuid": {
+                  "type": "keyword"
+                },
+                "version": {
+                  "type": "keyword"
+                }
+              }
+            },
+            "state": {
+              "properties": {
+                "beat": {
+                  "properties": {
+                    "name": {
+                      "type": "keyword"
+                    }
+                  }
+                },
+                "host": {
+                  "properties": {
+                    "architecture": {
+                      "type": "keyword"
+                    },
+                    "name": {
+                      "type": "keyword"
+                    },
+                    "os": {
+                      "properties": {
+                        "build": {
+                          "type": "keyword"
+                        },
+                        "family": {
+                          "type": "keyword"
+                        },
+                        "platform": {
+                          "type": "keyword"
+                        },
+                        "version": {
+                          "type": "keyword"
+                        }
+                      }
+                    }
+                  }
+                },
+                "input": {
+                  "properties": {
+                    "count": {
+                      "type": "long"
+                    },
+                    "names": {
+                      "type": "keyword"
+                    }
+                  }
+                },
+                "module": {
+                  "properties": {
+                    "count": {
+                      "type": "long"
+                    },
+                    "names": {
+                      "type": "keyword"
+                    }
+                  }
+                },
+                "output": {
+                  "properties": {
+                    "name": {
+                      "type": "keyword"
+                    }
+                  }
+                },
+                "service": {
+                  "properties": {
+                    "id": {
+                      "type": "keyword"
+                    },
+                    "name": {
+                      "type": "keyword"
+                    },
+                    "version": {
+                      "type": "keyword"
+                    }
+                  }
+                }
+              }
+            },
+            "timestamp": {
+              "format": "date_time",
+              "type": "date"
+            }
+          }
+        },
+        "beats_stats": {
```

Comment:
I wonder if we should roll APM server stuff into the Beats index. We already explicitly ignore `apm-server` data for all Beats monitoring UI, so it would help to avoid an extra template and set of indices, and the index is going to share a majority of the same fields with Beats.

Reply:
We've already run into a lot of confusion with using `beats` within `apm-server` - I think users who monitor their indices and don't use beats will be very surprised to see a beats index present.

Are there any security concerns that would push towards separating the indexes?

Reply:
Yeah, @ruflin, @graphaelli and I discussed this briefly off PR. The advantages I see of giving APM server its own set of monitoring indices are:

* We could, in the future, create a separate built-in role for APM server monitoring and restrict it's write access to only APM server monitoring indices.
* The mapping for APM server monitoring indices can grow to fit APM server needs without "polluting" the general beats mapping.
* (Relatively minor) It removes the filtering in the Monitoring UI code for excluding APM server documents when looking at Beats monitoring and, vice versa, for only including APM server documents when looking at APM server monitoring.

Reply:
The very big negative for adding a new index is that it creates a drain on Elasticsearch resources. Chances are, users won't have a lot of `apm-server` instances reporting to the Monitoring cluster, so creating a separate index for such a low scale use case means we're adding the shards for what will likely be about `24 * 60 * 60 / 10 = 8640` documents per day. At the same time, we'll require Elasticsearch to search each one of those indices for data whenever we try to summarize the stack views.

That's extremely wasteful, especially for small instances/clusters that can generally host monitoring data, but are now getting pushed closer to their limits because of the added load from threading the searches, joining the data in memory, growing the cluster state, and adding the segments.

> I think users who monitor their indices and don't use beats will be very surprised to see a beats index present.

We could get around this fear by either documenting it, or updating the index pattern for `.monitoring-beats-6-*` to be `.monitoring-beats-apm-6-*`. We could do this for 6.5 and add an alias to `.monitoring-beats-6-alias`, to the template, so that there is no impact to the UI.

> We could, in the future, create a separate built-in security role for APM server monitoring and restrict it's write access to only APM server monitoring indices.

We can still do this. The product user (Beats, Kibana, Logstash) writes indirectly to the monitoring data, so adding a user specific to APM does not require its own index. When Beats gains the ability to monitor APM then we can restrict access by adding a suffix to the index (e.g., `.monitoring-beats-apm-6-apm-*`) and limit the user to that, if we so choose. It would be ideal, even if we do split the indices, to think of them as a `.monitoring-*` index pattern rather than the subsets. This should help us, in the future, to avoid this kind of problem. We never should have separated Kibana into its own index. We did that for curation and this, but it's useless on its own, tiny, and never curated separately.

Reply:
I had a chat about this with @ycombinator last week and we expect that apm-server metrics potentially diverge from the Beats ones and we would prefer to keep it separate to have the flexibility in the future.

For the resource usage problem we should rather look into using ILM to have less indices and shards in my opinion instead of mixing multiple products into 1 index.

The other extreme we could go is having the data form all products in 1 index as we currently do with Metricbeat, but we should not start to mix both options.

Reply:
Hey @pickypg, I'd like to move this PR forward a bit this week. Could you please take a look at @ruflin's comment above? Thanks!

Reply:
@ruflin 

> we expect that apm-server metrics potentially diverge from the Beats ones and we would prefer to keep it separate to have the flexibility in the future.

Isolating data by the prefixed `type` has always worked for stack monitoring and nothing about that should be impacted by such deviations _unless_ they try to share a `type` (e.g., `beats_stats`) while changing the meaning of specific fields. I don't expect that to happen (versus augmenting it, which is fine).

> For the resource usage problem we should rather look into using ILM to have less indices and shards in my opinion instead of mixing multiple products into 1 index.

I agree that we should be using ILM/rollover when it is released, regardless of what we do here. However, at least for the short term, to avoid expanding the amount of data that we're storing on existing monitoring clusters, we will want to configure ILM to prune data based on time to match existing behavior (7 days by default). Longer term, particularly once Beats takes over, we can hopefully revisit this solution to be based more on size. Also, adding the shards and extra index mappings is an unnecessary impact that I really think that we should avoid.

> The other extreme we could go is having the data form all products in 1 index as we currently do with Metricbeat, but we should not start to mix both options.

I strongly think that we should move in that direction. With Beats, if we / the user decides that they want their data isolated by product, then it's as simple as modifying the index like how I am suggesting to add the `-apm` suffix. It really does boil down to a huge mistake that we aren't already doing this.

Think about the impact on the cluster from sharing the index relative to Beats-based usage:

- Searches use the same shards
- Bulk activity _can_ use the same shards

Versus separate indices:

- Searches use separate shards
- Index mappings are separated (growing the cluster state and routing table)
- Bulk activity cannot be optimized further than by product

Reply:
I chatted with @ruflin off-PR. For now, for the reasons that @pickypg has articulated in this thread, I'm going to **not** create a separate index (and template) for APM server monitoring documents. In the longer run, it's likely that we'll be changing indices altogether, perhaps consolidating all monitoring data into metricbeat indices; we can always revisit the indexing strategy then. So for now I'm going to update this PR to remove the index-related changes.

I will, however, keep the changes for a new built-in role and user for APM server monitoring, since asking users/admins to use beats users/roles could be easily confusing (as opposed to users/admins seeing monitoring indices with `beats` in the name as those are a bit more behind the scenes).